### PR TITLE
A model exporter for PyTorch

### PR DESCRIPTION
**Pull request by ezyang**
This is a big PR. I don't recommend looking at the commits individually, although you may find it useful to blame lines here and there. (Naming disclaimer: the project uses "toffee" in some places, but this will need to be renamed.)

A lot of folks contributed to this PR. Shout outs to @zdevito, @apaszke, @prigoyal, @killeent, @houseroad; as well as indirect contributions from @Yangqing, @bddppq and @jerryzh168.

**User facing interface.** At the moment, most functionality is made available from `torch.jit` and `torch.toffee` modules. Some refactoring is on going for these interfaces; `torch.jit` will be made into a private namespace and `torch.toffee` will grow end-to-end conversion functions.

**Key new pieces of code.**

* **The compiler IR.** There is a new, mutable, SSA, in-memory compiler IR representation which represents a computation in PyTorch. We maintain accurate def-use chains. The IR is defined in `torch/csrc/jit/ir.h`. Some unusual things about the IR:
  * Deep learning networks generally have multiple inputs AND outputs. However, computation nodes in our IR only have a single output; instead, we project out the particular output we are interested in using a Select node. The "Select invariant" states that any multiple return node (that's most nodes) always has exactly as many Select nodes as outputs. There are no first-class tuples in the language.
  * Computations in the graph are associated with stages, specifying whether or not they occurred in forwards/backwards/backwards-backwards. A computation in some stage cannot depend on a computation from a later stage; in fact, earlier stage computations will always be earlier in the IR than later stage computations.
  * Nodes in the graph are dynamically typed; while there are some special nodes (CppOp and PythonOp), most nodes are represented as strings with arbitrary dictionaries of attribute names to attribute values. Names are represented as interned strings in `torch/csrc/jit/interned_strings.h`; attribute dynamic dictionaries implemented in `torch/csrc/jit/attributes.h`.
  * There is a linter to enforce invariants. Run the linter early and often!
  * There are Python bindings using pybind11 for the IR in `torch/csrc/jit/python_ir.h`. These bindings are used in the definition of exporters for autograd functions.
* **The forward tracer.** The forward tracer instruments invocations to autograd Functions in the forward pass and records the execution into the compiler IR. 
  * We implemented this tracing logic twice: once for C++ functions, and once for Python functions. The key functions are `tracedApply` in `torch/csrc/autograd/function.cpp` (C++ side) and `trace_create` in `torch/csrc/autograd/python_function.cpp`. Each of these functions are traced as PythonOps and CppOps.
  * There is a calling convention change for C++ autograd functions; instead of writing `Transpose(0, 1).apply({ggI})[0]` you write `apply_fn<Transpose>(0, 1)(ggI)`.
  * The tracer can only trace what code was actually executed; e.g., it only sees the conditional branch that was taken.
  * Variables are marked as participating in a trace or not. A variable cannot participate in multiple traces simultaneously. You MUST dispose of a trace object (GC-wise) before you can use it again in another trace.
* **The Toffee exporter.** The Toffee exporter takes a raw trace from the forward tracer (consisting of PythonOps and CppOps) and converts it to a Toffee format IR, which can then be exported to protobufs. The code lives in `torch/csrc/toffee/export.cpp`. Feel free to ignore the logic inside the `passes_state_transparently` conditional; this is for handling backwards traces.
  * The way transformations are implemented is for each original op, we invoke the `primspec` associated with it (a new method associated with autograd Functions). This primspec takes as input the Toffee graph being constructed, and the inputs to the node (in the new Toffee IR graph), and extends the graph with the Toffee version of itself. Most primspecs are in Python, but C++ ops like BatchNorm and Convolution are done in C++.
  * After this transformation, we export it to protobufs using the nanopb library. We've checked in the nanopb-protoc generated files for Toffee IR in `torch/csrc/toffee.pb.h` and `cpp`; nanopb has a pretty user-unfriendly interface, so there is a small wrapper around it in `torch/csrc/toffee.h` which simulates the Google protobuf C++ interface. See 619e20dee57e9642a8357bdb5f23e51f8b88d05e how it works.
  * The manually checked in autogen files can be updated with `gen_toffee.sh`. If there are major changes to the protobuf, you may also have to update the C++ wrapper code.

**Experimental code.** There are some experimental pieces which are not exercised by the model exporter, but we'd like to merge in. I don't recommend spending too much time on these pieces of code; they are at high risk of being rewritten/being heavily refactored, and they shouldn't affect anything in PyTorch proper.

* **Optimization passes.** We have optimization passes for:
  * An "init" pass (soon to be removed; `torch/crsc/jit/init_pass.cpp`)
  * Dead code elimination (`torch/csrc/jit/dead_code_elimination.cpp`)
  * Operator fusion (`torch/csrc/jit/graph_fusion.cpp`). Once operators are fused into fusion groups, the fusion compiler (`torch/csrc/jit/fusion_compiler.cpp`) handles actually compiling this to CUDA. This has only been lightly tested and is expected to be broken at the moment.
* **The trace executor.** The trace executor takes a compiler IR trace, and converts it back into an autograd closure which we can then execute using the autograd engine. The main functionality for this is in `torch/crsc/autograd/functions/jit_closure.cpp`.
* **The backwards tracer.** The backwards tracer is responsible for intercepting invocations of backward functions which were traced, and recording their backward traces to the trace (so forwards and backwards can be cooptimized.) A good chunk of the functionality here is implemented in `torch/csrc/autograd/functions/special.cpp`
  * One significant source of complexity in the implementation is handling autograd functions whose backward passes are not traceable. In this case, we must fall back on the old PyTorch style (namely, generate an autograd trace when we compute forwards, and run that trace in backwards).

**Miscellaneous changes.** (In order GitHub renders files)

* New dependencies: pybind11 and nanopb. Pybind11 is a pretty nice wrapper for Python that can handle many conversions automatically. We intend keep migrating over PyTorch's Python wrapping code to use Pybind11 when applicable. Nanopb is a very small protobuf library; we use it to generate protobufs for the exporter. This prevents users from having to go through the pain of getting protobufs to work.
* Submodules. PyTorch repo now has submodules: gloo (pre-existing), pybind11 and nanopb. The model for when to subtree versus when to submodule is primarily a question of how likely we will need to apply local patches to a project in PyTorch, versus being able to go through the upstream cycle. There is also a technical reason why gloo was submoduled, see #2426
* Expect tests in the test suite. A new test method `assertExpected` lets you assert that a string matches some file which we have saved to disk. If the format changes, you can simply pass `--accept` to accept the new output. This makes it easy to see before and after in diffs.
* A new test suite for the JIT. While the JIT is not a public facing component in this diff, it is a key part of the export pipeline, and so tests for it are in test_jit.py
* There are C++ ports of chunk and split added to `torch/csrc/autograd/functions/tensor.cpp`
* `torch/csrc/autograd/engine.cpp` Autograd is now reentrant; you can call autograd backward from within a backwards function. The implementation is the diff to engine.cpp, and our strategy is to just call back into the scheduling loop so we continue to service requests for the worker thread even when we have kicked off a subgraph task.
* `torch/csrc/autograd/engine.cpp` Autograd callbacks have been generalized to support both pre and post callbacks. A pre callback is called before a function is invoked and can be used to rewrite inputs; a post callback is called after the function is invoked and we have output callbacks.
* New `TemplateEnv` class in `torch/csrc/jit/code_template.h` which can be used to format C++ code fragments. It supports key-based template variables and can handle indentation correctly.
* New utility `auto_unique_ptr` which constructs its object on the first dereference (rather than eagerly)
* New utility `ResourceGuard`, an RAII class which can be explicitly released within its lexical scope, or automatically released when the scope ends.
* New utility `fmap` which applies function to all elements of a vector (functionally, so no mutation)
* New parameter to `state_dict()`, `keep_vars`, which causes `state_dict` to return parameters as Variables rather than tensors. This is used internally by the tracer.

### CODE REVIEW

File: torch/csrc/toffee.h [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+#include "torch/csrc/toffee.pb.h"
+#include "torch/csrc/jit/assert.h"
+
+#include <pb_encode.h>
+#include <ATen/ATen.h>
+
+// #include <toffee/schema.h>
+// #include <google/protobuf/text_format.h>
+// #include <google/protobuf/io/zero_copy_stream_impl.h>
```
Comment (apaszke) [inline torch/csrc/toffee.h]: These should be removed

File: torch/csrc/jit/tracer.h [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+struct TracingState : public std::enable_shared_from_this<TracingState> {
+  TracingState()
+    : graph(new Graph())
+    , active(false) {}
+
+  std::shared_ptr<Graph> graph;
+  // void* is an unsafe TH.  NON-OWNING, so it might get invalidated.
+  // TODO: Perhaps, turn this into an owning reference.  The buffers
+  // are persistent, so this won't lead to a leak.
+  std::unordered_map<void*, Node*> buffer_map;
```
Comment (apaszke) [inline torch/csrc/jit/tracer.h]: Why does it need to be part of the tracing state? We could only give it to the exporter, right?

File: torch/csrc/toffee/export.cpp [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+  for (auto kv : old_buffer_map) {
+    buffer_map[kv.first] = envFn(kv.second);
+  }
+  ctx.buffer_map = &buffer_map;
+  // put the new outputs in our environment map, and
+  // copy the type from the input graph if they were not set by the
+  // primspec
+  auto setOutputs = [&](Node * node, const node_list & outputs) {
+    auto old_outputs = node->outputs();
+    JIT_ASSERTM(outputs.size() <= old_outputs.size(), "primspec produced too many outputs");
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: @ezyang Can you please add a comment saying when does it happen? I remember you once told me this, but I already forgot. Is this the case when exported IR should have some outputs that don't normally exit in PyTorch?

File: torch/csrc/toffee/export.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    size_t i = 0;
+    for(auto & old : old_outputs) {
+      // TODO: what if there are multiple trailing handle outputs?  That is
+      // a serious invariant violation...
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: Nothing. Add that to the linter

File: torch/csrc/toffee/export.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+      if(i >= outputs.size()) {
+        // primspecs do not deal with Handles at the moment
+        // so we map handles to Unused nodes
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: Except that Unused nodes no longer exist, so it would be good to change this comment

File: torch/jit.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    cpu_rng_state = torch.get_rng_state()
+    gpu_rng_state = None
+    if torch.cuda.is_available():
+        gpu_rng_state = torch.cuda.get_rng_state()
+
+    yield
+
+    torch.set_rng_state(cpu_rng_state)
+    if gpu_rng_state is not None:
+        torch.cuda.set_rng_state(gpu_rng_state)
```
Comment (apaszke) [inline torch/jit.py]: I think this only caches the RNG state of the current GPU.

File: test/common.py [hunk 7]
```diff
@@ -7,6 +7,8 @@
 from functools import wraps
 from itertools import product
 from copy import deepcopy
+import __main__
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/common.py]: This gives you the `__main__` module, i.e., the top level Python module which we initially entered; documented at https://docs.python.org/3/library/__main__.html. We use this in the "No expect file exists" to give a command line invocation which works out of the box.

File: test/expect/TestJit.test_batchnorm_export.expect [hunk 8]
```diff
@@ -0,0 +1,37 @@
+version: 1
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/expect/TestJit.test_batchnorm_export.expect]: How do you update the s3 when the format changes?

File: torch/autograd/_functions/basic_ops.py [hunk 9]
```diff
@@ -216,6 +229,12 @@ def backward(ctx, grad_output):
 class Negate(InplaceFunction):
 
     @staticmethod
+    def primspec(g, i, inplace=False):
+        if inplace:
+            return None
+        return g.appendNode(g.create("Scale", [i]).f_("scale", -1))
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/autograd/_functions/basic_ops.py]: @zdevito is working on it... someday :)

File: torch/csrc/jit/resource_guard.h [hunk 10]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+#include <functional>
+
+namespace torch { namespace jit {
+
+class ResourceGuard {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/jit/resource_guard.h]: It's used in `torch/csrc/jit/fusion_compiler.cpp`

File: torch/csrc/jit/tracer.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/jit/tracer.h]: The best time to initialize the buffer_map is when we initiate the trace (since we get in the buffers from the state_dict.) But then, the most convenient way to pass this along to the exporter is to put it in the tracing state.

File: torch/csrc/toffee/export.cpp [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+#include "torch/csrc/autograd/functions/convolution.h"
+#include "torch/csrc/jit/dead_code_elimination.h"
+#include "torch/csrc/utils/functional.h"
+#include <ATen/ATen.h>
+
+#include <fstream>
+#undef NDEBUG
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#define NDEBUG
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/toffee/export.cpp]: CC @zdevito

File: torch/csrc/toffee/export.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/toffee/export.cpp]: It's a `<=` and not an `==` because we don't want to force primspecs to handle implicit Handle returns. Writing the comment.

File: torch/nn/_functions/thnn/auto.py [hunk 12]
```diff
@@ -27,6 +28,11 @@ def _make_function_class_criterion(class_name, update_output, update_grad_input,
         additional_arg_idx += 1
 
     @staticmethod
+    def primspec(*args, **kwargs):
+        a = primspec_fn(*args, **kwargs)
+        return a
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/nn/_functions/thnn/auto.py]: No you can't, because primspec_fn is a variable lexically brought into scope by the enclosing function definition.

File: torch/nn/_functions/thnn/auto.py [hunk 13]
```diff
@@ -122,6 +129,10 @@ def _initialize_buffers(ctx, fn_name):
         return tuple(additional_args)
 
     @staticmethod
+    def primspec(*args, **kwargs):
+        return primspec_fn(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/nn/_functions/thnn/auto.py]: Ditto.

File: test/expect/TestJit.test_batchnorm_export.expect (hunk 8 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply test/expect/TestJit.test_batchnorm_export.expect]: upload a new file?

[the remark this answers is not in the document]
Comment (apaszke) [reply test/expect/TestJit.test_batchnorm_export.expect]: although it would be nicer to use a linter

File: torch/csrc/toffee/export.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/toffee/export.cpp]: Remove it. It was to try to enable some better debugging, but it never ended up working.

**Comment by ezyang** [comment]
Latest batch of commits from 'jit' branch are cdecfae..d618864.

File: torch/jit.py (hunk 6 above)
Comment (ezyang) [reply torch/jit.py]: Indeed, this seems to be the case. Should we add a torch.cuda helper for setting all of the states? (Maybe set_rng_state should set everything, and there should be a set_current_gpu_rng_state...)

File: docs/source/toffee.rst [hunk 14]
```diff
@@ -2,6 +2,113 @@ torch.toffee
...
 
+Example: End-to-end AlexNet from PyTorch to Caffe2
+--------------------------------------------------
+
+Here is a simple script which exports a pretrained AlexNet as defined in
+torchvision into Toffee IR.  It runs a single round of inference and then
+saves the resulting traced model to ``alexnet.proto``.  (We recommend
+running this inference on GPU, because PyTorch does not have an efficient
+CPU convolution implementation.  It may take some time to initialize the
+CUDA instance.)::
```
Comment (apaszke) [inline docs/source/toffee.rst]: Let's not put this in the docs

File: torch/nn/_functions/thnn/activation.py [hunk 15]
```diff
@@ -8,6 +8,11 @@
 
 
 class PReLU(Function):
+
+    @staticmethod
+    def primspec(g, input, weight):
+        return g.appendNode(g.create("PRelu", [input, weight]))
```
Comment (apaszke) [inline torch/nn/_functions/thnn/activation.py]: Note that PReLU can work in two modes in PyTorch (weights can be shared between all channels), but it might not be true in Caffe

**Comment by ezyang** [comment]
@soumith When CI turns green this is ready to merge.

**Review by apaszke — APPROVED** [review]
